View: Block protocol viewer

jonasstrehle opened the pull request:
Closes #13

Commit:
Created block protocol viewer (issue#13) branch from initial setup

Commit:
Intialized example DatexBlockProtocolViewer to understand project structure and router functionality

Commit:
started implementing grid layout

trying to implement a responsive but nicely aligned grid layout that will display the Datex Blocks with hovering for more info

Commit:
loading bin and json files works on a basic level, commiting as a backup safety

Commit:
loading bin and json files works on a basic level, commiting as a backup safety

Commit:
fetching example datex block protocol binary data to use for reference for further layout improvements

Commit:
fetching example datex block protocol binary data to use for reference for further layout improvements

Commit:
unstyled but working recursive display Datex Blocks to set a minimal working start

Commit:
unstyled but working recursive display Datex Blocks to set a minimal working start

Commit:
rendered hex text can now be highlighted and copied + some codestructure and variable names changed

Commit:
rendered hex text can now be highlighted and copied + some codestructure and variable names changed

Commit:
not dynamic but statically working implementation to handle fields that habe subFields (e.g. Flags) to give different hover info or display hex numbers seperated

Commit:
not dynamic but statically working implementation to handle fields that habe subFields (e.g. Flags) to give different hover info or display hex numbers seperated

Commit:
latest stable flexbox implementation before redoing the layout with display grid

Commit:
latest stable flexbox implementation before redoing the layout with display grid

Commit:
implemented basic accordion structure for each section to be able to collapse unwanted block fields

Commit:
implemented basic accordion structure for each section to be able to collapse unwanted block fields

Commit:
minimum working prototype for grid layout,
DatexBlockField.vue became redundant

Commit:
minimum working prototype for grid layout,
DatexBlockField.vue became redundant

Commit:
cleaned up unused functions, comments and css selectors for better readability

Commit:
cleaned up unused functions, comments and css selectors for better readability

Commit:
added byte cutoff for values that are too long and some code cleanup

Commit:
added byte cutoff for values that are too long and some code cleanup

jonasstrehle committed:
Rebase to main

jonasstrehle committed:
prettier

jonasstrehle committed:
prettier

Commit:
fixed general coloring issues because I wasn't able to load var colors from the main.css

Commit:
fixed general coloring issues because I wasn't able to load var colors from the main.css

Commit:
made coloring depend on category of field provided by speck repo

Commit:
made coloring depend on category of field provided by speck repo

Commit:
fix #38

Commit:
fix #38

Commit:
added back calculating the column width in a seperate function
this value is not accessible with repeat(auto-fit, ..) but I need to manipulate n:th children in a dynamic grid

Commit:
added back calculating the column width in a seperate function
this value is not accessible with repeat(auto-fit, ..) but I need to manipulate n:th children in a dynamic grid

Commit:
prettier

Commit:
prettier

Commit:
split fields into new component

Commit:
split fields into new component

Commit:
added tailwind wherever it seemed reasonable

Commit:
added tailwind wherever it seemed reasonable

Commit:
prettier

Commit:
prettier

Commit:
reverted tailwind that messed up the padding

Commit:
reverted tailwind that messed up the padding

Commit:
actually fixed the category coloring issue and started field hovering feature

Commit:
actually fixed the category coloring issue and started field hovering feature

Commit:
fix #39

Commit:
fix #39

Commit:
added primitive expand/collapse support for fields

Commit:
added primitive expand/collapse support for fields

Commit:
fix #40

Commit:
fix #40

Commit:
prettier and temporary scroll bar

Commit:
prettier and temporary scroll bar

Commit:
improved fix #39
the dom structure is worse now but the ui behaves more like desired now

Commit:
improved fix #39
the dom structure is worse now but the ui behaves more like desired now

Commit:
prettier

Commit:
prettier

Commit:
fix #41

Commit:
fix #41

Commit:
prettier

Commit:
prettier

Commit:
updated subFields seperation functionality to work with latest speck version of 0.0.11

Commit:
updated subFields seperation functionality to work with latest speck version of 0.0.11

Commit:
solution idea for desired hover functionality in issue #42

Commit:
solution idea for desired hover functionality in issue #42

Commit:
starting to use the correct types instead of just generic objects

Commit:
starting to use the correct types instead of just generic objects

Commit:
almost correct use of specific types

Commit:
almost correct use of specific types

Commit:
almost correct use of types & prettier

Commit:
almost correct use of types & prettier

Commit:
new component for info view and a bunch of prettier css

Commit:
new component for info view and a bunch of prettier css

Commit:
prettier

Commit:
prettier

Commit:
turned BlockInfo into shadcn table and defined upcoming features

Commit:
turned BlockInfo into shadcn table and defined upcoming features

Commit:
reworked Info Table layout but the close button functionality is broken

Commit:
reworked Info Table layout but the close button functionality is broken

Commit:
beginning of restructuring of all files to improve the way I can handle data binding and states

Commit:
beginning of restructuring of all files to improve the way I can handle data binding and states

Commit:
correctly mapped category from FieldDefinition to color of field

Commit:
correctly mapped category from FieldDefinition to color of field

Commit:
added cariable for selected Field so that the Bytes View and the InfoView can share the variable in the same state

Commit:
added cariable for selected Field so that the Bytes View and the InfoView can share the variable in the same state

Commit:
restarting the BlockProtocolInfo

Commit:
restarting the BlockProtocolInfo

Commit:
added specific type and restarted the InfoView with new type

Commit:
added specific type and restarted the InfoView with new type

Commit:
added colorMapping and improved css hover conditions when fields are selected for the InfoView.

Commit:
added colorMapping and improved css hover conditions when fields are selected for the InfoView.

Commit:
very ugly and way too nested implementation of subFields

Commit:
very ugly and way too nested implementation of subFields

Commit:
a little prettier with outsorced code in ts functions

Commit:
a little prettier with outsorced code in ts functions

Commit:
new component fieldWrapper that splits the code for better readability

Commit:
new component fieldWrapper that splits the code for better readability

Commit:
prettier

Commit:
prettier

Commit:
obsolote files from old, unstructured implementation

Commit:
obsolote files from old, unstructured implementation

Commit:
prettier

Commit:
prettier

Commit:
had to move functions and declarations to different files to get the desired effect when expanding or collapsing a field
hovering has to be redone because of new structure

Commit:
had to move functions and declarations to different files to get the desired effect when expanding or collapsing a field
hovering has to be redone because of new structure

Commit:
when selecting a field, it is now correctly colored in while all other fields are greyed out

Commit:
when selecting a field, it is now correctly colored in while all other fields are greyed out

Commit:
basic hovering to grey out other fields

Commit:
basic hovering to grey out other fields

Commit:
small cleanup

Commit:
small cleanup

Commit:
improved hover

Commit:
improved hover

Commit:
changed info closing x button to a lucide icon

Commit:
changed info closing x button to a lucide icon

Commit:
moved some elements in info for better organization

Commit:
moved some elements in info for better organization

jonasstrehle commented on src/components/BlockFieldWrapper.vue:
```diff
@@ -0,0 +1,70 @@
+
+const emit = defineEmits(['field-clicked']);
+const handleClick = () => {
+  // console.log(props.field);
```
Remove logs

jonasstrehle commented on src/components/BlockViewer/BlockFieldWrapper.vue:
```diff
@@ -0,0 +1,70 @@
+  props.sectionId == props.selectedField?.sectionIndex &&
+  props.fieldId == props.selectedField.fieldIndex;
+
+const bytesCutoff: number = 25;
```
Put to a settings / globals file for the component or move to the top of the file and add a explaining comment on what it does and if there are some limits to the range.

jonasstrehle commented on src/components/BlockSection.vue:
```diff
@@ -0,0 +1,46 @@
+        class="contents"
+        :class="`${
+          selectedField &&
+          sectionId == selectedField.sectionIndex &&
```
Also use the === equals op

jonasstrehle commented on src/views/BlockProtocolInfoView.vue:
```diff
@@ -0,0 +1,75 @@
+const field = computed(
+  () => props.structure[props.selectedField.sectionIndex]?.fields[props.selectedField.fieldIndex],
+);
+console.log(field);
```
Remove log

jonasstrehle commented on src/views/BlockViewer/BlockProtocolInfoView.vue:
```diff
@@ -0,0 +1,75 @@
+      </div>
+      <Separator />
+      <p v-if="'parsedValue' in field" class="py-2 text-sm">
+        Value: {{ String(field.parsedValue) }}
```
Maybe we can add a mapping / handler for displaying the parsed value in another way, such as showing the bytes of the magic number not as js array, but also as hex or whatever. Like if there is a big string maybe we'll decide to cut it off later and only expand it on click?! So I suggest to put a renderParsedValue method into place what we can extend later.

jonasstrehle commented on src/views/BlockProtocolInfoView.vue:
```diff
@@ -0,0 +1,75 @@
+          <TableRow>
+            <TableHead class="w-[30%]">name</TableHead>
+            <TableHead class="w-[30%]">id</TableHead>
+            <TableHead class="w-[40%]">parsedValue</TableHead>
```
I would name the column header to `value` instead of `parseValue`. Also is there a benefit in also showing the id field?

jonasstrehle commented on src/views/BlockProtocolInfoView.vue:
```diff
@@ -0,0 +1,75 @@
+        <TableBody>
+          <TableRow v-for="(subField, i) in field.subFields" :key="i">
+            <TableCell>{{ subField.name }}</TableCell>
+            <TableCell>{{ 'id' in subField ? subField.id : '-' }}</TableCell>
```
Maybe we can show the `-` when the field is empty in a grayed out manner, or display `no entry` / `none` in italic and other color?

jonasstrehle commented on src/views/DatexBlockProtocolView.vue:
```diff
@@ -0,0 +1,70 @@
+  selectedField.value = data;
+};
+
+const handleInfoCloseButtionClick = () => {
```
spelling: buttion -> button

jonasstrehle commented on src/views/BlockViewer/DatexBlockProtocolView.vue:
```diff
@@ -0,0 +1,70 @@
+  </div>
+</template>
+
+<!-- DONE don't seperate subfields when the field is not expanded -->
```
would be cool to see this in an issue and not in the source code here, you might also put it to a TODO.md file if you plan to have it in the history (either project wide or in the views.

jonasstrehle commented on src/types/BlockViewer/FieldIdentifier.ts:
I'd suggest to move the block viewer and all dependencies (ProtocolBytesView, types, ...) to a separate folder.
```
src/views/
    BlockViewer/
        types/
    BlockProtocolBytesView.vue
    ...
```

or 
```
src/views/
    BlockViewer/
src/types/
    BlockViewer/
        blockProtocolView.ts
    ...
```

Also I'd suggest to change the file name of the ts-file to be all camel case instead of using hypens, just for consistency reasons.

jonasstrehle commented on src/views/BlockViewer/DatexBlockProtocolView.vue:
```diff
@@ -0,0 +1,70 @@
+    'https://raw.githubusercontent.com/unyt-org/datex-specification/refs/heads/main/assets/structures/dxb.json',
+  )
+).json();
+const blockDataExample: Uint8Array = new Uint8Array(
```
Can we add the block Uint8Array as prop to the component, so we can construct different components by passing the prop accordingly?

benStre commented on src/components/BlockViewer/BlockField.vue:
```diff
@@ -0,0 +1,73 @@
+  fieldDef: FieldDefinition | undefined;
+}>();
+
+const uint8ToHexString = (b: number): string => b.toString(16).padStart(2, '0');
```
maybe byteToHexString would be better

jonasstrehle commented on src/components/BlockViewer/BlockField.vue:
```diff
@@ -0,0 +1,73 @@
+
+const uint8ToHexString = (b: number): string => b.toString(16).padStart(2, '0');
+
+const categories = ['purple', 'red', 'yellow', 'green', 'blue', 'dark_blue'];
```
Rename to caps `CATEGORIES`

benStre commented on src/components/BlockViewer/BlockField.vue:
```diff
@@ -0,0 +1,73 @@
+const uint8ToHexString = (b: number): string => b.toString(16).padStart(2, '0');
+
+const categories = ['purple', 'red', 'yellow', 'green', 'blue', 'dark_blue'];
+const getColor = (s: string | undefined): string => {
```
we should probably use normal functions instead of arrow functions to make the distinction between values und functions clearer

benStre commented on src/components/BlockViewer/BlockField.vue:
```diff
@@ -0,0 +1,73 @@
+const getColor = (s: string | undefined): string => {
+  if (!s) return 'var(--chart-1)';
+
+  const index = categories.findIndex((color) => s === color);
```
indexOf would also do the job here instead of findIndex

benStre commented on src/components/BlockFieldWrapper.vue:
```diff
@@ -0,0 +1,70 @@
+};
+
+const fieldIsSelectedField = () =>
+  props.sectionId == props.selectedField?.sectionIndex &&
```
you should always use `===` for strict type checking

benStre commented on src/components/BlockViewer/BlockFieldWrapper.vue:
```diff
@@ -0,0 +1,70 @@
+const displaySubfields = (): boolean =>
+  'subFields' in props.field &&
+  props.field.subFields.reduce(
+    (acc: number, subField: ParsedField) => acc + subField.bytes.length,
```
I wouldn't use explicit parameter types for callback closure when the types are already inferred from the usage

benStre replied:
I would also convert `displaySubfields` to a computed value instead of a function

benStre commented on src/components/BlockViewer/BlockFieldWrapper.vue:
```diff
@@ -0,0 +1,70 @@
+    </div>
+    <div v-else class="subfield-wrapper contents">
+      <BlockField
+        v-for="(subField, index) in 'subFields' in field ? field.subFields : []"
```
I would add parentheses here to make it more readable, it's a bit confusing with the two `in` operators

lennartstoelting replied:
I used to have it like this but the way prettier is set up, formatting the code removes these parentheses
I could rewrite it in a helper function but it seems to me like we don't gain enough from this

benStre replied:
ah okay thats weird - maybe we can change the prettier rules for this in the future, but then we can resolve this for now.

benStre commented on src/components/BlockViewer/BlockSection.vue:
```diff
@@ -0,0 +1,46 @@
+        v-for="(field, index) in section.fields"
+        :key="index"
+        class="contents"
+        :class="`${
```
would use an object here instead of a template string: 
```tsx
{'selected-field': ...}
```

benStre commented on src/views/BlockViewer/DatexBlockProtocolView.vue:
```diff
@@ -0,0 +1,70 @@
+
+
+// this will later not be fetched from the example data
+const jsonDataExample: StructureDefinition = await (
```
rename this variable to dxbDefinition or something like this

benStre commented on src/views/BlockViewer/BlockProtocolBytesView.vue:
```diff
@@ -0,0 +1,112 @@
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+const props = defineProps<{
+  structure: ParsedStructure;
+  structureDef: StructureDefinition | undefined;
```
structureDef must always exist as a prop here and can never be undefined

benStre commented on src/views/BlockProtocolBytesView.vue:
```diff
@@ -0,0 +1,112 @@
+        <BlockSection
+          v-if="section.fields.length > 0"
+          :section="section"
+          :sectionDef="structureDef?.sections.find((sect) => sect.name == section.name)"
```
sectionDef must never be undefined - so if find does not returning anything, you can throw an error and make sure it is always defined when passed into BlockSection. You could outsource this in a helper function

benStre commented on src/components/BlockViewer/BlockSection.vue:
```diff
@@ -0,0 +1,46 @@
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+const props = defineProps<{
+  section: ParsedSection;
+  sectionDef: SectionDefinition | undefined;
```
sectionDef must never be undefined. This should also be changed everywhere the sectionDef variable is used, so the `?.` operator can be removed

Commit:
removed logs

Commit:
removed logs

Commit:
strict type checking

Commit:
strict type checking

Commit:
removed id column from subFields table in InfoView

Commit:
removed id column from subFields table in InfoView

Commit:
moved block viewer and all dependencies to a seperate folder

Commit:
moved block viewer and all dependencies to a seperate folder

Commit:
mapping variable renamed to caps

Commit:
function renamed

Commit:
changed arrow functions to normal functions for better distinction

Commit:
more reanimg of functions for clearer distinction

Commit:
used indexOf instead of findIndex

Commit:
removed explicit parameter types for callback closure, the types are already inferred from the usage

Commit:
use an object for class binding instead of template string

Commit:
renamed variable

Commit:
spelling errors resolved

Commit:
introduced settings.ts to set global variables for component

Commit:
made showing Id in the subfield info a togglable setting

Commit:
grey out dash when subfield has no id

Commit:
turned inline todo comments into project issues

Commit:
wrapped the whole view in a seperate parent file to fetch the block.bin with.
This will ensure that when calling the component in the future, it works with any binary file, not just the example.

Commit:
sectionDef must never be undefined, created helper function to catch error

Commit:
more helper functions to catch case when there is no definition for a given field

Commit:
displaySubfields is now a computed value instead of function

Commit:
started helper function renderParsedValue
this will later on be used to display the parsed Value correctly according to what type it is or how long the value is

Commit:
feat: dynamically create category mapping

categories of fields used to be statically defined, now we use the names from the definition imported from the datex-specifications repo

Commit:
feat: use lucide icons for Theme switcher

refers to #45

Commit:
feat: hovering a byte shows a tooltip of its name

beginning of #42

Commit:
fix: cleanup for type FieldIdentifier

Commit:
feat: implement color in infoview

two different drafts for possible styling of infoView

start of issue #47

benStre commented on src/views/BlockViewer/DatexBlockProtocolView.vue:
```diff
@@ -0,0 +1,44 @@
+        @bytes-section-field-clicked="handleBytesSectionFieldClick"
+      ></BlockProtocolBytes>
+    </div>
+    <div class="m-1 max-h-2/5 overflow-y-auto rounded-lg border bg-red-600" v-if="selectedField">
```
Is this still WIP right now? otherwise, can you remove the bg-red-600?

lennartstoelting replied:
yes sorry the bg-red-600 was work in progress to see where I want to incorporate the color in the InfoView, visually connecting the info view and the bytes view. It is now in the correct spot, showing the color that corresponds to the selected fields category

benStre reviewed (changes requested):
Looks good! I tested it locally again, and noticed that at least on my large screen, the font size for the byte blocks is very small with 100% zoom (~11px). Can you take a look at this?
And can you also rebase the branch on the latest state of release/0.0.1, there are currently some conflicts that prevent this PR from being merged.

Commit:
fix: padding in infoView

Commit:
fix: visually cleaned up InfoView Layout

incorporated category color, more visually distinct spacing and the table prioritizes the name column width when shrinking

Commit:
fix: updated package-lock json

Commit:
fix: changed FieldIdentifier type definition

Commit:
adjusted linehieght for bytes
